[RDY] Migrate legacy test 83.

lucc opened the pull request:
I have two open questions about this test:
1. Is iconv always available in nvim or should the test check it with `has()`?  See comment in the test.
2. I split this PR into two commits (will squash after discussion).  278e015 migrates the test and it runs.  But the test uses some data files for the tag search and some tags files.  In the old test these are extracted from test.in or saved under git in the test directory. 92d28ac removes all of these files and writes them to disk at runtime.  They are stored as literal strings in the lua test file.  _Which of the two commits does The Right Thing?_  (I prefer 92d28ac.)

lucc committed:
tests: Migrate legacy test 83.

fwalch commented on test/functional/legacy/083_tag_search_with_file_encoding_spec.lua:
```diff
@@ -0,0 +1,217 @@
+      'abc100	Xtags3.txt	/\x82`\x82a\x82b\n'
+      )
+
+    -- TODO what about the availability of iconv in neovim?
```
iconv is not always available, unfortunately... not sure how to get the results of `has('iconv')` into the tests, but maybe you can add an `if` at the beginning and just return if it iconv is not available.

justinmk replied:
Agree with @fwalch except instead of just `return`ing, call `pending()` so that the *total* test count stays the same, but we know which tests were skipped. E.g.:

```lua
if not has_iconv then
  pending('skipped (missing iconv)')
else
  it('is working', function()
  end)
end
```

See also: https://github.com/neovim/neovim/wiki/Unit-tests#guidelines-for-writing-tests

lucc committed:
tests: Improve test 83 and add `pending()` call.

If nvim was compiled without `has("iconv")` this test is skipped.

lucc commented:
rebased and updated

@justinmk I did not find any reference to `has_iconv` which you mentioned so I wrote a local function. But I am quite unhappy with the result. It is slow and I can not reverse it (see TODO in the code).

lucc commented:
ref for a better overview: #1286

lucc commented:
rebased to use #2932.

lucc commented:
I now consider this PR fixed. If somebody reviews it again I will squash and mark it RDY.

justinmk commented on test/functional/legacy/083_tag_search_with_file_encoding_spec.lua:
```diff
@@ -0,0 +1,110 @@
+local has_iconv = function()
+  return os.execute(helpers.nvim_prog .. ' -u NONE -i NONE --embed --cmd ' ..
+    [['if has("iconv")|q|else|cq|endif']]) == 0
+end
```
A bit nicer:

```lua
local function has_iconv()
  clear()  -- ensures session
  return 1 == helpers.eval('has("iconv2")')
end
```

justinmk replied:
At note about your comment from an older commit:

> If I reverse the check ( ... !has("iconv") ... ) I do not get a pending message 

That's because `0` is not false in lua. Only `false` and `nil` are false in lua. Sorry for your trouble.

lucc replied:
I think you mean `has("iconv")`, I could not find the string "iconv2" anywhere in the source.

The old comment was removed in the commit that fixed the issue. The problem was that I did not pass an empty function to `pending` on the second arg.

justinmk commented on test/functional/legacy/083_tag_search_with_file_encoding_spec.lua:
```diff
@@ -0,0 +1,110 @@
+  end)
+
+  if not has_iconv() then
+    pending('skipped (missing iconv)', function() end)
```
`missing iconv` is enough, because the test runner will say "skipped ..."

justinmk commented:
@lucc Glad to see this. LGTM after `has_iconv()` simplification + squash.

lucc commented:
RDY
